### PR TITLE
Context providers and consumers should bailout on already finished work

### PR DESCRIPTION
**Pull request by acdlite**
Fixes bug where a consumer would re-render even if its props and context had not changed.

- [x] Confirm https://github.com/facebook/react/issues/12218 is fixed
- [x] Confirm https://github.com/facebook/react/issues/12246 is fixed
- [x] Integrate context fuzz tester into main fuzz tester, since that would have caught this bug. These tests are most effective when you combine as many features together as possible.

### CODE REVIEW

**Comment by acdlite** [comment]
@gaearon Ready for review

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -921,11 +918,19 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
 
-    if (changedBits !== 0) {
+    if (changedBits === 0) {
+      if (hasLegacyContextChanged()) {
+        // Normally we can bail out on props equality but if context has changed
+        // we don't do the bailout and we have to reuse existing props instead.
+      } else if (oldProps === newProps) {
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
+    } else if (changedBits !== 0) {
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Could just be `else`

**Comment by bvaughn** [comment]
This change seems to make sense to me. 👍

**Comment by acdlite** [comment]
@sebmarkbage I'm gonna merge this as-is so I can start the sync. If you have additional feedback, I'll fix in a follow up.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -977,12 +977,13 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
...
     const render = newProps.children;
 
-    if (typeof render !== 'function') {
-      invariant(
-        false,
-        "A context consumer was rendered with multiple children, or a child that isn't a function. " +
-          'A context consumer expects a single child that is a function. ' +
-          'If you did pass a function, make sure there is no trailing or leading whitespace around it.',
+    if (__DEV__) {
+      warning(
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Should we move this to context creation then? Or do we already warn there?

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: Context creation is too early. Needs to check on every render. Like a prop type.

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: I mean `createElement` when we know type is a context consumer.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: Ah I see. Idk, not sure what the advantages are. I thought about using `propTypes` but if the plan is to remove those eventually? Although maybe this is an indication that would shouldn't...

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: >Idk, not sure what the advantages are. 

You get a JS stack trace pointing where it happened.

**Pull request merged**